client/src/main/scala/drt/client/services/SPACircuit.scala

```diff
@@ -100,17 +127,51 @@ class CrunchStateHandler[M](pointInTime: ModelR[M, Option[SDateLike]], modelRW:
 
       effectOnly(allEffects)
 
-    case GetCrunchStateAfter(delay)
-    =>
-      log.info(s"Re-requesting flights")
+    case GetCrunchUpdatesAfter(delay) =>
       val getCrunchStateAfterDelay = Effect(Future(GetCrunchState())).after(delay)
-      effectOnly(getCrunchStateAfterDelay)
+      effectOnly(getCrunchStateAfterDelay + Effect(Future(HideLoader())))
+
+    case UpdateCrunchStateFromUpdates(crunchUpdates) =>
+      log.info(s"Client got ${crunchUpdates.flights.size} flights & ${crunchUpdates.minutes.size} minutes from CrunchUpdates")
+
+      val someStateExists = value._1.isReady
+
+      val newState = if (someStateExists) {
+        val existingState = value._1.get
+        updateStateFromUpdates(crunchUpdates, existingState)
+      } else newStateFromUpdates(crunchUpdates)
+
+      updated((Ready(newState), crunchUpdates.latest), Effect(Future(HideLoader())))
+  }
+
+  def newStateFromUpdates(crunchUpdates: CrunchUpdates): CrunchState = {
+    CrunchState(0L, 0, crunchUpdates.flights, crunchUpdates.minutes)
+  }
 
-    case UpdateCrunchState(crunchState)
-    =>
-      log.info(s"client got ${crunchState.flights.size} flights (from CrunchState)")
+  def updateStateFromUpdates(crunchUpdates: CrunchUpdates, existingState: CrunchState): CrunchState = {
+    val lastMidnightMillis = SDate.midnightThisMorning().millisSinceEpoch
+    val flights = updateAndTrimFlights(crunchUpdates, existingState, lastMidnightMillis)
+    val minutes = updateAndTrimMinutes(crunchUpdates, existingState, lastMidnightMillis)
+    CrunchState(flights = flights, crunchMinutes = minutes, crunchFirstMinuteMillis = 0L, numberOfMinutes = 0)
+  }
 
-      updated(Ready(crunchState), Effect(Future(HideLoader())))
+  def updateAndTrimMinutes(crunchUpdates: CrunchUpdates, existingState: CrunchState, lastMidnightMillis: MillisSinceEpoch): Set[Crunch.CrunchMinute] = {
```

Comment:
should we move the boundaries of the day (e.g. last midnight) into a single global location so we can eventually change? Perhaps not one for this PR.

Reply:
Sure, I'll delay it for now